File: rust/src/creational/decorator.rs
trait DataSource {
    fn write(&mut self, data: &str);
    fn read(&self) -> String;
}

struct FileDataSource {
    storage: String,
}
impl FileDataSource {
    fn new() -> Self {
        Self {
            storage: String::new(),
        }
    }
}
impl DataSource for FileDataSource {
    fn write(&mut self, data: &str) {
        self.storage = data.to_string();
    }
    fn read(&self) -> String {
        self.storage.clone()
    }
}

struct EncryptionDecorator {
    inner: Box<dyn DataSource>,
}
impl EncryptionDecorator {
    fn new(inner: Box<dyn DataSource>) -> Self {
        Self { inner }
    }
}
impl DataSource for EncryptionDecorator {
    fn write(&mut self, data: &str) {
        self.inner.write(&format!("[enc]{}", data));
    }
    fn read(&self) -> String {
        self.inner.read().replace("[enc]", "")
    }
}

struct CompressionDecorator {
    inner: Box<dyn DataSource>,
}
impl CompressionDecorator {
    fn new(inner: Box<dyn DataSource>) -> Self {
        Self { inner }
    }
}
impl DataSource for CompressionDecorator {
    fn write(&mut self, data: &str) {
        self.inner.write(&format!("[zip]{}", data));
    }
    fn read(&self) -> String {
        self.inner.read().replace("[zip]", "")
    }
}

// Builder-стиль
struct DataSourceBuilder {
    current: Box<dyn DataSource>,
}
impl DataSourceBuilder {
    fn new(base: Box<dyn DataSource>) -> Self {
        Self { current: base }
    }

    fn add_encryption(mut self) -> Self {
        self.current = Box::new(EncryptionDecorator::new(self.current));
        self
    }

    fn add_compression(mut self) -> Self {
        self.current = Box::new(CompressionDecorator::new(self.current));
        self
    }

    fn build(self) -> Box<dyn DataSource> {
        self.current
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example() {
        let mut ds = DataSourceBuilder::new(Box::new(FileDataSource::new()))
            .add_compression()
            .add_encryption()
            .build();

        ds.write("Hello");
        println!("{}", ds.read());
    }
}

File: java/src/main/java/structural/Decorator.java
package structural;

interface DataSource {

    void writeData(String data);

    String readData();
}

class FileDataSource implements DataSource {

    private final String filename;

    public FileDataSource(String filename) {
        this.filename = filename;
    }

    @Override
    public void writeData(String data) {
        System.out.println("File " + filename + ": " + data);
    }

    @Override
    public String readData() {
        return "Reading from file " + filename;
    }
}

// Base decorator
abstract class DataSourceDecorator implements DataSource {

    protected DataSource wrappee;

    public DataSourceDecorator(DataSource wrappee) {
        this.wrappee = wrappee;
    }

    @Override
    public void writeData(String data) {
        wrappee.writeData(data);
    }

    @Override
    public String readData() {
        return wrappee.readData();
    }
}

class EncryptionDecorator extends DataSourceDecorator {

    public EncryptionDecorator(DataSource wrappee) {
        super(wrappee);
    }

    @Override
    public void writeData(String data) {
        super.writeData("[encrypted]" + data);
    }

    @Override
    public String readData() {
        return "[decrypted]" + super.readData();
    }
}

class CompressionDecorator extends DataSourceDecorator {

    public CompressionDecorator(DataSource wrappee) {
        super(wrappee);
    }

    @Override
    public void writeData(String data) {
        super.writeData("[compressed]" + data);
    }

    @Override
    public String readData() {
        return "[uncompressed]" + super.readData();
    }
}

// Builder helper that chains decorators together
class DataSourceBuilder {

    private DataSource current;

    public DataSourceBuilder(DataSource base) {
        this.current = base;
    }

    public DataSourceBuilder addEncryption() {
        current = new EncryptionDecorator(current);
        return this;
    }

    public DataSourceBuilder addCompression() {
        current = new CompressionDecorator(current);
        return this;
    }

    public DataSource build() {
        return current;
    }
}

@SuppressWarnings("unused")
final class DecoratorExample {

    private DecoratorExample() {
    }

    public static void example() {
        DataSource dataSource = new DataSourceBuilder(new FileDataSource("data.txt"))
                .addCompression()
                .addEncryption()
                .build();

        dataSource.writeData("Hello World");
        System.out.println(dataSource.readData());
    }
}
